src/services/codefixes/inferFromUsage.ts

```diff
@@ -665,29 +670,116 @@ namespace ts.codefix {
             }
         }
 
-        function getTypeFromUsageContext(usageContext: UsageContext, checker: TypeChecker): Type | undefined {
-            if (usageContext.isNumberOrString && !usageContext.isNumber && !usageContext.isString) {
-                return checker.getUnionType([checker.getNumberType(), checker.getStringType()]);
+        const inferrers: ((k: UsageContext, checker: TypeChecker) => Type[])[] = [
+            inferFromPrimitive,
+            inferFromContext,
+            inferFromObject,
+            inferFromStructure,
+        ];
+
+        function getTypeFromUsageContext(k: UsageContext, checker: TypeChecker, fallback: Type): Type {
+            const inferences = [fallback];
+            for (const inferrer of inferrers) {
+                for (const t of inferrer(k, checker)) {
+                    inferences.push(t);
+                    updateInferenceProperties(t, k, checker);
+                }
             }
-            else if (usageContext.isNumber) {
-                return checker.getNumberType();
+            const anons = inferences.filter(i => checker.getObjectFlags(i) & ObjectFlags.Anonymous) as AnonymousType[];
+            const good = [];
+            if (!k.hasNonVacuousNonAnonymousType && anons.length) {
+                good.push(unifyAnonymousTypes(anons, checker));
             }
-            else if (usageContext.isString) {
-                return checker.getStringType();
+            good.push(...inferences.filter(i => !isVacuousInference(i, k, checker)));
+            return checker.getWidenedType(checker.getUnionType(good));
+        }
+
+        function unifyAnonymousTypes(anons: AnonymousType[], checker: TypeChecker) {
+            if (anons.length === 0) {
+                return checker.getNeverType();

```

Comment:
I think this is unreachable because that was checked before calling this function.